### ISSUE TITLE
Inheritance errors

### ISSUE THREAD

**Issue by ygModesto** [mapstruct/mapstruct#1885]
Hi, I created a project to show 2 problems. https://github.com/ygModesto/hierarchy

The `issue1` branch shows an error in the generated mapper (DepartmentMapperImpl).
Mapstruct tries to assign a `Collection<UserTO>` when it should assign a `Collection<UserBaseTO>`

To try to solve the `issue1` I create the code of the `issue2` branch but it generates another error.

Mapstruct notifies that it finds Ambiguous mappings, I can solve it with some workaround but in projects with more classes and inheritances it reaches an impossible point.

**Comment by filiphr** [comment]
Thanks for the reproducal. So to summarise the problems:

### Issue1

If there is a setter that accepts `Collection<Base>` and a mapper that returns `Collection<Implementation>` we are going to assign directly to the base (which is not correct).

In order to circumvent this you can use an adder and the `CollectionMappingStrategy.ADDER_PREFERRED`.

### Issue2

That one is a bit more tricky. First ambiguous mapping is because there are 2 methods that could be used to create the `Collection<UserBaseTO> userToUserBaseTO(Collection<User> users);`. You could remove the `userToUserBaseTO` method. However, then in `DepartmentMapper` there is a problem that we pick both collection methods for mapping the collection. This problem is tricky and is similar to the second one.

Currently I am not sure how we can fix those. Is adding an adder option for you?

**Comment by Zegveld** [comment]
@ygModesto , I'm closing this issue as the main issue (issue1) seems to be resolved with 1.5.0.Beta2.

It has no compilation errors and the generated code looks fine.
The generated code for mapping to the `Collection<UserBaseTO>`:
```
    protected Collection<UserBaseTO> userCollectionToUserBaseTOCollection(Collection<User> collection) {
        if ( collection == null ) {
            return null;
        }

        Collection<UserBaseTO> collection1 = new ArrayList<UserBaseTO>( collection.size() );
        for ( User user : collection ) {
            collection1.add( userMapper.userToUserTO( user ) );
        }

        return collection1;
    }
```

**Comment by filiphr** [comment]
Thanks for validating this @Zegveld
